From 74ef555267f83e6e0e365112749dd8b743ee2b87 Mon Sep 17 00:00:00 2001
From: n8fr8 <nathan@freitas.net>
Date: Mon, 19 Sep 2016 16:04:04 -0400
Subject: [PATCH 1/1] add limits.h to support MAX_SIZE on Android

---
 src/ext/trunnel/trunnel.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ext/trunnel/trunnel.c b/src/ext/trunnel/trunnel.c
index 0ed75aa..dbb9476 100644
--- a/src/ext/trunnel/trunnel.c
+++ b/src/ext/trunnel/trunnel.c
@@ -13,6 +13,11 @@
 #include <stdlib.h>
 #include <string.h>
 #include "trunnel-impl.h"
+ /* Get SIZE_MAX.  */
+#ifdef __BIONIC__
+#include <limits.h>
+#endif
+
 
 #if defined(__BYTE_ORDER__) && defined(__ORDER_LITTLE_ENDIAN__) && \
 	__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
-- 
1.9.1

