commit d42eec7683ce9d47fc846e511e815ec1b2950c27
Author: n8fr8 <nathan@freitas.net>
Date:   Fri May 4 17:32:51 2012 -0400

    remove hardening for android

diff --git a/configure.ac b/configure.ac
index facd3c6..ad405ea 100644
--- a/configure.ac
+++ b/configure.ac
@@ -24,7 +24,7 @@ AC_CHECK_HEADERS([sys/types.h])
 AC_CHECK_HEADERS([netinet/in.h])
 
 ### Compiler Hardening ###
-AX_ENABLE_HARDENING
+##AX_ENABLE_HARDENING
 
 ### Libraries ###
 
